### PR TITLE
preserve value infos if they are needed

### PR DESCRIPTION
**Pull request by anderspapitto**
in particular, we'll need this so that we can actually store inferred
shapes without them vanishing

### CODE REVIEW

File: onnx/common/ir_pb_converter.cc [hunk 1]
```diff
@@ -468,6 +468,9 @@ void encodeGraph(ONNX_NAMESPACE::GraphProto * p_g, const std::shared_ptr<Graph>
     ONNX_NAMESPACE::ValueInfoProto* v = p_g->add_output();
     encodeValueInfo(v, output);
   }
+
+  std::set<Value*> all_outputs(g->outputs().begin(), g->outputs().end());
```
Comment (bddppq) [inline onnx/common/ir_pb_converter.cc]: unordered_set

File: onnx/test/optimizer_test.py [hunk 2]
```diff
@@ -10,10 +10,10 @@
 
 class TestOptimizer(unittest.TestCase):
 
-    def _optimized(self, graph, name):
+    def _optimized(self, graph, opts):
         orig_model = helper.make_model(graph, producer_name='onnx-test')
         orig_model_str = orig_model.SerializeToString()
-        optimized_model_str = onnx.optimizer.optimize(orig_model_str, [name])
+        optimized_model_str = onnx.optimizer.optimize(orig_model_str, list(opts))
```
Comment (bddppq) [inline onnx/test/optimizer_test.py]: why add a `list` call here?

Comment (anderspapitto) [reply onnx/test/optimizer_test.py]: ++ not needed, will remove

File: onnx/common/ir_pb_converter.cc (hunk 1 above)
Comment (anderspapitto) [reply onnx/common/ir_pb_converter.cc]: ++ will switch

Comment (houseroad) [reply onnx/common/ir_pb_converter.cc]: Call it graph_outputs?

File: onnx/common/ir_pb_converter.cc [hunk 3]
```diff
@@ -483,6 +486,17 @@ void encodeGraph(ONNX_NAMESPACE::GraphProto * p_g, const std::shared_ptr<Graph>
     }
     for(auto output : node->outputs()) {
       p_n->add_output(value_name(output));
+
+      // only save it if it has content
```
Comment (houseroad) [inline onnx/common/ir_pb_converter.cc]: To be more clear here, we should say, the output information (including shape/type) is already stored in graph's output.

File: onnx/test/optimizer_test.py [hunk 4]
```diff
@@ -93,11 +93,29 @@ def test_fuse_transpose_into_gemm(self):
...
+            "test",
+            [helper.make_tensor_value_info("X", TensorProto.FLOAT, (2, 3, 4))],
+            [helper.make_tensor_value_info("A", TensorProto.FLOAT, (4, 3, 2))])
+        graph.value_info.extend(
+            [helper.make_tensor_value_info("Y", TensorProto.FLOAT, (3, 2, 4))]
+        )
+
+        optimized_model = self._optimized(graph, [])
+
+        assert len(optimized_model.graph.value_info) == 1
```
Comment (houseroad) [inline onnx/test/optimizer_test.py]: You may also want to check value_info's name, type and shape.

File: onnx/test/optimizer_test.py [hunk 5]
```diff
@@ -93,11 +93,29 @@ def test_fuse_transpose_into_gemm(self):
...
+        graph = helper.make_graph(
+            [trans1, trans2, trans3],
+            "test",
+            [helper.make_tensor_value_info("X", TensorProto.FLOAT, (2, 3, 4))],
+            [helper.make_tensor_value_info("A", TensorProto.FLOAT, (4, 3, 2))])
+        graph.value_info.extend(
+            [helper.make_tensor_value_info("Y", TensorProto.FLOAT, (3, 2, 4))]
+        )
+
+        optimized_model = self._optimized(graph, [])
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply onnx/test/optimizer_test.py]: sure ++

File: onnx/common/ir_pb_converter.cc (hunk 3 above)
Comment (anderspapitto) [reply onnx/common/ir_pb_converter.cc]: ++

File: onnx/test/optimizer_test.py (hunk 4 above)
Comment (anderspapitto) [reply onnx/test/optimizer_test.py]: sure ++

**Pull request merged**